File: Complementar - Estruturas de dados (Didático)/Algoritmos35/Alg35.cpp
/*
♦ Algoritmo 35: Estrutura de Dados Não Linear Dinâmica (não Estática) de Árvore Binária: Árvore Rubro-Negra.
*/


#include <iostream>

template<typename T>
class No {
public:
    T valor;
    No<T>* esquerda;
    No<T>* direita;
    bool vermelho;

    No(T valor) : valor(valor), esquerda(nullptr), direita(nullptr), vermelho(true) {}
};

template<typename T>
class ArvoreRubroNegra {
private:
    No<T>* raiz;
    static const bool VERMELHO = true;
    static const bool PRETO = false;

    bool isVermelho(No<T>* no) {
        if (no == nullptr) return false;
        return no->vermelho == VERMELHO;
    }

    No<T>* rotacaoEsquerda(No<T>* no) {
        No<T>* temp = no->direita;
        no->direita = temp->esquerda;
        temp->esquerda = no;
        temp->vermelho = no->vermelho;
        no->vermelho = VERMELHO;
        return temp;
    }

    No<T>* rotacaoDireita(No<T>* no) {
        No<T>* temp = no->esquerda;
        no->esquerda = temp->direita;
        temp->direita = no;
        temp->vermelho = no->vermelho;
        no->vermelho = VERMELHO;
        return temp;
    }

    void inverterCores(No<T>* no) {
        no->vermelho = !no->vermelho;
        no->esquerda->vermelho = !no->esquerda->vermelho;
        no->direita->vermelho = !no->direita->vermelho;
    }

    No<T>* inserirRecursivo(No<T>* no, T valor) {
        if (no == nullptr) return new No<T>(valor);

        if (valor < no->valor)
            no->esquerda = inserirRecursivo(no->esquerda, valor);
        else if (valor > no->valor)
            no->direita = inserirRecursivo(no->direita, valor);
        else
            return no;

        if (isVermelho(no->direita) && !isVermelho(no->esquerda))
            no = rotacaoEsquerda(no);
        if (isVermelho(no->esquerda) && isVermelho(no->esquerda->esquerda))
            no = rotacaoDireita(no);
        if (isVermelho(no->esquerda) && isVermelho(no->direita))
            inverterCores(no);

        return no;
    }

    void imprimirEmOrdemRecursivo(No<T>* no) {
        if (no != nullptr) {
            imprimirEmOrdemRecursivo(no->esquerda);
            std::cout << no->valor << " ";
            imprimirEmOrdemRecursivo(no->direita);
        }
    }

public:
    ArvoreRubroNegra() : raiz(nullptr) {}

    void inserir(T valor) {
        raiz = inserirRecursivo(raiz, valor);
        if (raiz != nullptr) raiz->vermelho = PRETO;
    }

    void imprimirEmOrdem() {
        imprimirEmOrdemRecursivo(raiz);
    }

    bool estaVazia() {
        return raiz == nullptr;
    }
};

int main() {
    ArvoreRubroNegra<int> arvore;
    arvore.inserir(10);
    arvore.inserir(5);
    arvore.inserir(20);
    arvore.inserir(3);
    arvore.inserir(8);
    arvore.inserir(15);

    std::cout << "Em Ordem:" << std::endl;
    arvore.imprimirEmOrdem();

    return 0;
}

File: Estruturas de Dados e Complexidade de Algoritmos/Algoritmos35/Alg35.py
'''
♦ Algoritmo 35: Estrutura de Dados Não Linear Dinâmica (não Estática) de Árvore Binária: Árvore Rubro-Negra.
'''


class No:
    def __init__(self, valor):
        self.valor = valor
        self.esquerda = None
        self.direita = None
        self.vermelho = True  # Por padrão, os nós são vermelhos quando adicionados à árvore

class ArvoreRubroNegra:
    def __init__(self):
        self.raiz = None
        self.VERMELHO = True
        self.PRETO = False

    # Verifica se um nó é vermelho
    def isVermelho(self, no):
        if no is None:
            return False
        return no.vermelho == self.VERMELHO

    # Realiza uma rotação para a esquerda em torno do nó dado
    def rotacaoEsquerda(self, no):
        temp = no.direita
        no.direita = temp.esquerda
        temp.esquerda = no
        temp.vermelho = no.vermelho
        no.vermelho = self.VERMELHO
        return temp

    # Realiza uma rotação para a direita em torno do nó dado
    def rotacaoDireita(self, no):
        temp = no.esquerda
        no.esquerda = temp.direita
        temp.direita = no
        temp.vermelho = no.vermelho
        no.vermelho = self.VERMELHO
        return temp

    # Inverte as cores do nó e de seus filhos
    def inverterCores(self, no):
        no.vermelho = not no.vermelho
        no.esquerda.vermelho = not no.esquerda.vermelho
        no.direita.vermelho = not no.direita.vermelho

    # Insere um novo nó na árvore Rubro-Negra
    def inserir(self, valor):
        self.raiz = self.inserirRecursivo(self.raiz, valor)
        self.raiz.vermelho = self.PRETO  # A raiz sempre será preta

    # Método auxiliar recursivo para inserção
    def inserirRecursivo(self, no, valor):
        if no is None:
            return No(valor)

        # Inserção normal como em uma árvore de busca binária
        if valor < no.valor:
            no.esquerda = self.inserirRecursivo(no.esquerda, valor)
        elif valor > no.valor:
            no.direita = self.inserirRecursivo(no.direita, valor)
        else:
            return no  # Valor já existe na árvore

        # Correção da propriedade de cor após a inserção
        if self.isVermelho(no.direita) and not self.isVermelho(no.esquerda):
            no = self.rotacaoEsquerda(no)
        if self.isVermelho(no.esquerda) and self.isVermelho(no.esquerda.esquerda):
            no = self.rotacaoDireita(no)
        if self.isVermelho(no.esquerda) and self.isVermelho(no.direita):
            self.inverterCores(no)

        return no

    # Método para imprimir a árvore em ordem
    def imprimirEmOrdem(self):
        self.imprimirEmOrdemRecursivo(self.raiz)

    # Método auxiliar recursivo para impressão em ordem
    def imprimirEmOrdemRecursivo(self, no):
        if no is not None:
            self.imprimirEmOrdemRecursivo(no.esquerda)
            print(no.valor, end=" ")
            self.imprimirEmOrdemRecursivo(no.direita)

    # Método para verificar se a árvore Rubro-Negra está vazia
    def estaVazia(self):
        return self.raiz is None

# Teste
arvore = ArvoreRubroNegra()
arvore.inserir(10)
arvore.inserir(5)
arvore.inserir(20)
arvore.inserir(3)
arvore.inserir(8)
arvore.inserir(15)

print("Em Ordem:", end=" ")
arvore.imprimirEmOrdem()
